ml_tools/preprocess.py

```diff
@@ -0,0 +1,332 @@
+
+            # apply level and contrast shift
+            level_adjust = random.normalvariate(0, LEVEL_OFFSET)
+            contrast_adjust = tools.random_log(0.9, (1 / 0.9))
+
+            data[:, 0] *= contrast_adjust
+            data[:, 0] += level_adjust
+            data[:, 1] *= contrast_adjust
+
+        if random.random() <= 0.50:
+            # when we flip the frame remember to flip the horizontal velocity as well
+            data = np.flip(data, axis=3)
+            data[:, 2] = -data[:, 2]
+    return data
+
+
+def reisze_cv(image, dim, interpolation=cv2.INTER_LINEAR, extra_h=0, extra_v=0):
+    return cv2.resize(
+        image,
+        dsize=(dim[0] + extra_h, dim[1] + extra_v),
+        interpolation=interpolation,
+    )
+
+
+def preprocess_frame(
+    data, output_dim, use_thermal=True, augment=False, preprocess_fn=None
+):
+    if use_thermal:
+        channel = TrackChannels.thermal
+    else:
+        channel = TrackChannels.filtered
+    data = data[channel]
+
+    max = np.amax(data)
+    min = np.amin(data)
+    if max == min:
+        return None
+
+    data -= min
+    data = data / (max - min)
+    np.clip(data, a_min=0, a_max=None, out=data)
+
+    data = data[np.newaxis, :]
+    data = np.transpose(data, (1, 2, 0))
+    data = np.repeat(data, output_dim[2], axis=2)
+    data = reisze_cv(data, output_dim)
+
+    # preprocess expects values in range 0-255
+    if preprocess_fn:
+        data = data * 255
+        data = preprocess_fn(data)
+    return data
+
+
+def preprocess_movement(
+    data,
+    segment,
+    square_width,
+    regions,
+    channel,
+    preprocess_fn=None,
+    augment=False,
+):
+    segment = preprocess_segment(
+        segment, augment=augment, filter_to_delta=False, default_inset=0
+    )
+
+    segment = segment[:, channel]
+
+    # as long as one frame it's fine
+    square, success = square_clip(segment, square_width, (FRAME_SIZE, FRAME_SIZE), type)
+    if not success:
+        return None
+    dots, overlay = movement_images(
+        data,
+        regions,
+        dim=square.shape,
+        channel=channel,
+        require_movement=True,
+    )
+    dots = dots / 255
+    overlay, success = normalize(overlay, min=0)
+    if not success:
+        return None
+
+    data = np.empty((square.shape[0], square.shape[1], 3))
+    data[:, :, 0] = square
+    data[:, :, 1] = dots  # dots
+    data[:, :, 2] = overlay  # overlay
+
+    if preprocess_fn:
+        for i, frame in enumerate(data):
+            frame = frame * 255
+            data[i] = preprocess_fn(frame)
+    return data
+
+
+def normalize(data, min=None, max=None, new_max=1):
+    if max is None:
+        max = np.amax(data)
+    if min is None:
+        min = np.amin(data)
+    if max == min:
+        if max == 0:
+            return np.zeros((data.shape)), False
+        return data / max, False
+    data -= min
+    data = data / (max - min) * new_max
+    return data, True
+
+
+def movement_images(
```

Comment:
This is quite confusing having both the overlay and the dots code together.   Maybe separate it into two functions?  Then we can explore models with one or other or both the extra movement images.   We don't want to be caulculating something that doesn't give us much extra value.